### ISSUE TITLE
trailing_stop_positive_offset used even when trailing_stop_only_offset_is_reached is False

### ISSUE THREAD

**Issue by holycacao** [freqtrade/freqtrade#6828]
## Describe your environment

  * Operating system: Ubuntu 20.04.4 LTS
  * Python Version: Python 3.9.9
  * CCXT version: ccxt==1.80.61
  * Freqtrade Version: freqtrade 2022.4.1
  
## Your question

I _think_ this may be a bug, but I'm probably just not understanding the code regarding trailing stop loss handling when `trailing_only_offset_is_reached` is `False`. I would expect the behavior to be if the current profit is greater than `trailing_stop_positive` (not offset), adjust the stop loss value accordingly.

However, it looks like the code is still requiring the current profit be greater than `trailing_stop_positive_offset`. Here's the block of code from line 996 to 1012 in `freqtrade/strategy/interface.py` (my comments prefixed with `HC`):
```py
        if self.trailing_stop and (sl_lower_long or sl_higher_short):
            # HC: offset initialized to trailing stop positive offset
            sl_offset = self.trailing_stop_positive_offset

            # Make sure current_profit is calculated using high for backtesting.
            bound = low if trade.is_short else high
            bound_profit = current_profit if not bound else trade.calc_profit_ratio(bound)

            # HC: here the trailing stop positive offset is ignored if trailing only offset is false (as expected) 
            if not (self.trailing_only_offset_is_reached and bound_profit < sl_offset):
                # HC: but here, the profit must be greater than the  trailing stop positive offset -- not trailing stop positive!
                if self.trailing_stop_positive is not None and bound_profit > sl_offset:
                    stop_loss_value = self.trailing_stop_positive
                    logger.debug(f"{trade.pair} - Using positive stoploss: {stop_loss_value} "
                                 f"offset: {sl_offset:.4g} profit: {current_profit:.2%}")

                trade.adjust_stop_loss(bound or current_rate, stop_loss_value)
```

**Comment by xmatthias** [comment]
You'll need to be careful with indentation - maybe.

what changes with the if statement is the assignment `stop_loss_value = self.trailing_stop_positive` - which is skipped if profit is < offset.

If that's not the case, then stoploss will be adjusted according to the regular stop value (instead of `trailing_stop_positive`).

it might be interpretation, too - I've seen `trailing_only_offset_is_reached` interpreted wrongly before.

`trailing_only_offset_is_reached` doesn't mean "ignore offset" - instead, it means "keep `stoploss` static UNTIL offset is reached`.
So the stoploss is static until you reach the offset - and then starts trailing.

`trailing_stop_positive` on the other hand defines a different stoploss (which might be applied once the offset is reached).

so you could have a -10% stoploss until the offset of 5% is reached (either trailing from the start - or trail only once 5% is reached - keeping the initial stop at -10% from the entry point - depending on `trailing_only_offset_is_reached`) - and then trail with a -2% stoploss.

**Comment by holycacao** [comment]
Let's take the [Trailing stop loss, custom positive loss](https://www.freqtrade.io/en/stable/stoploss/#trailing-stop-loss-custom-positive-loss) example from the docs and walk through the code together. The example states using the following config:
```
    stoploss = -0.10
    trailing_stop = True
    trailing_stop_positive = 0.02
```
Notice `trailing_stop_positive_offset` is not in the example config. Now, let me walk through the entire `stop_loss_reached` function using the example from the docs (bot buys an asset at $100) and annotate it (my comments prefixed with `HC`):
```py
    def stop_loss_reached(self, current_rate: float, trade: Trade,
                          current_time: datetime, current_profit: float,
                          force_stoploss: float, low: float = None,
                          high: float = None) -> ExitCheckTuple:
        # HC: Initial stop loss value set to stoploss from config (-0.10)
        stop_loss_value = force_stoploss if force_stoploss else self.stoploss

        # HC: First time this func is called after the trade is made, this will set the stop loss
        #     to $90 (assume a $0 fee for simplicity). Afterwards, do nothing since initial=True.
        trade.adjust_stop_loss(trade.open_rate, stop_loss_value, initial=True)

        # HC: Ignoring this and the next if statement since not using a custom stoploss func
        dir_correct = (trade.stop_loss < (low or current_rate)
                       if not trade.is_short else
                       trade.stop_loss > (high or current_rate)
                       )

        if self.use_custom_stoploss and dir_correct:
            stop_loss_value = strategy_safe_wrapper(self.custom_stoploss, default_retval=None
                                                    )(pair=trade.pair, trade=trade,
                                                      current_time=current_time,
                                                      current_rate=current_rate,
                                                      current_profit=current_profit)
            # Sanity check - error cases will return None
            if stop_loss_value:
                # logger.info(f"{trade.pair} {stop_loss_value=} {current_profit=}")
                trade.adjust_stop_loss(current_rate, stop_loss_value)
            else:
                logger.warning("CustomStoploss function did not return valid stoploss")

        # HC: From here on, let's assume the price is now $102. $102 > $90 so sl_lower_long is True
        sl_lower_long = (trade.stop_loss < (low or current_rate) and not trade.is_short)
        sl_higher_short = (trade.stop_loss > (high or current_rate) and trade.is_short)

        # HC: Both the trailing_stop config and sl_lower_long are True so enter this block
        if self.trailing_stop and (sl_lower_long or sl_higher_short):
            # HC: sl_offset = ?? This is the first problem with the example from the docs. It does
            #     not mention that trailing_stop_positive_offset is required if using
            #     trailing_stop_positive and must be greater than trailing_stop_positive.
            #     I think this is a bug. The offset should only be required when
            #     trailing_only_offset_is_reached is True. It sounds like we disagree on this - at
            #     the very least, the docs should be updated to explain this.
            #     So for now, let's assume sl_offset = 0.03
            sl_offset = self.trailing_stop_positive_offset

            bound = low if trade.is_short else high
            # HC: Price is $102 so bound_profit is 0.02
            bound_profit = current_profit if not bound else trade.calc_profit_ratio(bound)

            # HC: trailing_only_offset_is_reached is False, so enter this block
            if not (self.trailing_only_offset_is_reached and bound_profit < sl_offset):
                # HC: trailing_stop_positive is set, but 0.02 is not greater than 0.03: skip!
                # In the docs example, it states the stop loss would now be $99.96, but this is not true!
                # The stop loss would not kick in until the profit is greater than the offset!
                # I think this comparison should be bound_profit > self.trailing_stop_positive.
                # If the user specified to only trail when the offset is reached, we verfied that
                # in the above if stmt.
                if self.trailing_stop_positive is not None and bound_profit > sl_offset:
                    stop_loss_value = self.trailing_stop_positive
                    logger.debug(f"{trade.pair} - Using positive stoploss: {stop_loss_value} "
                                 f"offset: {sl_offset:.4g} profit: {current_profit:.2%}")

                # adjust stop loss to -10% even though we're out of the red and maybe even greater
                # than trailing_stop_positive :(
                trade.adjust_stop_loss(bound or current_rate, stop_loss_value)

        # HC: The rest of this deals with trade exit logic
```
To me, the `trailing_stop_positive_offset` should _only_ be used to make the initial switch over from static stop loss to trailing stop loss when `trailing_only_offset_is_reached` is `True`. But if not, `trailing_stop_positive` should trigger the switch. What I'm suggesting is this:
```py
            if not (self.trailing_only_offset_is_reached and bound_profit < sl_offset):
                # Specific handling for trailing_stop_positive
                if self.trailing_stop_positive is not None and bound_profit > self.trailing_stop_positive:
                    stop_loss_value = self.trailing_stop_positive
                    logger.debug(f"{trade.pair} - Using positive stoploss: {stop_loss_value} "
                                 f"offset: {sl_offset:.4g} profit: {current_profit:.2%}")

                trade.adjust_stop_loss(bound or current_rate, stop_loss_value)
```
Again, even if you don't agree with this proposed logic, at the very least, the docs need to be updated to include how `trailing_stop_positive_offset` is used in the **Trailing stop loss, custom positive loss** case.

**Comment by xmatthias** [comment]
`trailing_stop_positive_offset` defaults to 0. It's therefore not relevant for this part of the docs, nor is it required in the config/strategy (it defaults to 0). 
every profit > 0 is seen as "positive" unless you specify it differently. 
You can obviously change this to use 1% for a different stoploss - but that doesn't invalidate the documentation.

I think [this part](https://www.freqtrade.io/en/stable/stoploss/#trailing-stop-loss-only-once-the-trade-has-reached-a-certain-offset) explains it quite well. The part you linked does NOT cover or include `trailing_only_offset_is_reached` - but describes behavior without that option.

> If `"trailing_only_offset_is_reached": true` then the trailing stoploss is only activated once the offset is reached. Until then, the stoploss remains at the configured stoploss. This option can be used with or without trailing_stop_positive, but uses trailing_stop_positive_offset as offset.

---

your proposed logic is clearly wrong and would only start trailing once the stoploss value is reached (using the trailing value as offset).
You'd therefore enforce that a 2% trailing positive stoploss would start at 2% profit. The current solution allows to trail once 4% (or whatever arbitrary number) is reached - allowing you to avoid a "trailing" scenario with a tight stoploss when a pair only "swings positive" for a moment - to pull back moments later.


i'd suggest to read the full stoploss documentation to fully understand the logic in place.
Your descriptions / understandings suggest that you ONLY read the parts you're interested in - but not the parts that explain the logic you find buggy (see link above).

**Comment by holycacao** [comment]
You're right that the code I proposed would enforce that a 2% trailing positive stop loss would start at 2% profit. This is because I thought that `trailing_stop_positive_offset` **had** to be greater than `trailing_stop_positive`, which is true except for the 0 case (like you mentioned). My mistake.

I did read the documentation - I think it does a good job of explaining the scenarios, but could be more clear around how exactly `trailing_stop_positive`, `trailing_stop_positive_offset`, and `trailing_only_offset_is_reached` interact. Specifically, how `trailing_stop_positive_offset` independently affects trailing stop behavior. Consider the following incrementally additive configs and their trailing stop behavior (please correct me if I'm wrong):

```
stoploss = -0.10
trailing_stop = True
```
10% stop loss trailing profit

```
stoploss = -0.10
trailing_stop = True
trailing_stop_positive = 0.02
```
10% static stop loss until profit > 0, then 2% trailing profit

```
stoploss = -0.10
trailing_stop = True
trailing_stop_positive = 0.02
trailing_stop_positive_offset = 0.03
```
10% static stop loss until profit > 3%, then 2% trailing profit. This scenario is what I think could help clarify how `trailing_stop_positive_offset` independently affects trailing stop.

```
stoploss = -0.10
trailing_stop = True
trailing_stop_positive = 0.02
trailing_stop_positive_offset = 0.03
trailing_only_offset_is_reached = True
```
10% stop loss trailing profit until profit > 3%, then 2% trailing profit

I think introducing `trailing_stop_positive_offset` in the **Trailing stop loss, custom positive loss** section would help clarify how it works. It's implicitly covered in that section because it defaults to 0, but it would be helpful to note that it must be 0 or greater than `trailing_stop_positive` and to provide an additional example with it non-zero showing how it affects the stop loss. I'm more than happy to submit a PR for this if you would like.

**Comment by xmatthias** [comment]
> I'm more than happy to submit a PR for this if you would like.

If you think the documentation could be improved to be clearer, then i'm more than happy to receive a PR for it 💯 
For me it's obvious - but i wrote (or reviewed) both the code as well as the docs - which  sometimes makes it difficult to see where it's unclear (because for me, with my preexisting knowledge - it's crystal clear).

**Comment by Vaadasch** [comment]
Sorry to hijack this 2-years old thread, @xmatthias, but it seems to me that trailing_only_offset_is_reached is useless.
Pardon my abruptness, it is for clarity. This single thing doesn't remove the quality of all the code, sometime we only need another pair of eyes 🙂

`trailing_only_offset_is_reached` need `trailing_stop_positive_offset` to be set.
And `trailing_stop_positive_offset`, when set, only set new `stoploss` (or `trailing_stop_positive` if set) when offset is reached.
Either with `trailing_only_offset_is_reached = True` or not

So we could have `trailing_only_offset_is_reached = (trailing_stop_positive_offset is not None)` and that's it, no?
And implicitely, in the code, it seems that it is already there.

```python 
if not (self.trailing_only_offset_is_reached and bound_profit < sl_offset): 
```
Can be translated to : 
```python 
if ( not self.trailing_only_offset_is_reached ) or (not bound_profit < sl_offset) 
```
equivalent to : 
```python 
if  not self.trailing_only_offset_is_reached or bound_profit > sl_offset
```
So whatever the value of `self.trailing_only_offset_is_reached`, if `bound_profit > sl_offset`, the condition is executed
(i suppose the check if sl_offset is not None is implicit in the comparison by raising an exception?)
When `self.trailing_only_offset_is_reached = False`,  the condition is executed whatever bound_profit > or < of sl_offset
That is the default behavior if we are not setting any sl_offset.

What I suggest :
```python
           if sl_offset is None or bound_profit > sl_offset:
                # Specific handling for trailing_stop_positive
                if self.trailing_stop_positive is not None:
                    stop_loss_value = self.trailing_stop_positive
                    logger.debug(
                        f"{trade.pair} - Using positive stoploss: {stop_loss_value} "
                        f"offset: {sl_offset:.4g} profit: {bound_profit:.2%}"
                    )

                trade.adjust_stop_loss(bound or current_rate, stop_loss_value)
```

Either way, when we are setting a `trailing_stop_positive_offset`, that is because we want a `trailing_only_offset_is_reached = True`, because we would not had setting it in the first place.

Am I completely wrong here?
I would be happy to do a PR if you confirm these assertions.

**Comment by xmatthias** [comment]
> Either way, when we are setting a trailing_stop_positive_offset, that is because we want a trailing_only_offset_is_reached = True, because we would not had setting it in the first place.

Well - no.

start with the most enhanced example of trailing stop (the [one including all properties](https://www.freqtrade.io/en/stable/stoploss/#trailing-stop-loss-only-once-the-trade-has-reached-a-certain-offset)) without going into custom_stoploss territory.

```
    stoploss = -0.10
    trailing_stop = True
    trailing_stop_positive = 0.02
    trailing_stop_positive_offset = 0.03
    trailing_only_offset_is_reached = True
```

`trailing_stop =True` (only this) means it'll trail. so at 1% profit - stoploss will be set to -9% (from open).
If we now include `trailing_stop_positive` - it'll trail at a different rate depending on profit (in the above case - it would trail at -10% until 3% profit - and then at a 2% offset to high.

now adding `trailing_stop_positive_offset =True` - this means - stoploss will remain at -10% (from open) until profit has reached `stop_profit_offset` - and will THEN trail at 2%.

If `trailing_stop_positive_offset =False` however - trailing will start immediately.

changing this example values to be a bit clearer:

```
    stoploss = -0.10
    trailing_stop = True
    trailing_stop_positive = 0.03
    trailing_stop_positive_offset = 0.15
    trailing_only_offset_is_reached = True
```
This configuration will mean
* keep stoploss at -10% until 15% profit is reached (so at 11% profit - stoploss will be at 10% below the trade's open rate).
* Once 15% profit is reached, start trailing at -3%


now with `trailing_only_offset_is_reached` - the one thing that changes is what happens between "open" and "15% profit".
If it's true - stoploss will stay at -10% and will not trail until 15% profit is reached
if it's false - stoploss will start trailing immediately - so at 1% profit, stop will be at -9% from open (and so forth).

In a very simplified form - think about it in the following way (orange is the price/profit - red is with "trailing_only_offset_is_reached =True" - blue is what changes if `trailing_only_offset_is_reached =False`.

![Image](https://github.com/user-attachments/assets/39352e69-a83f-4bf5-b226-010a103cccb5)


If you think it's unclear in the (current) docs - happy to see a PR improving how this is explained
but from a behavior perspective - i don't really see what would have to change.

**Comment by Vaadasch** [comment]
@xmatthias, perfect, thanks a lot.
>blue is what changes if `trailing_only_offset_is_reached =False`

That is the default value of `trailing_only_offset_is_reached` that was missing for me.

The example in the docs suggest to set it to True, but from what you explained, that is already the default behavior (I mean, default when we use the `trailing_stop_positive_offset`) => FALSE, i didn't re-read the doc when i've wrote this
So what is interesting with that parameter `trailing_only_offset_is_reached` is when we set it to False ?

**Comment by xmatthias** [comment]
well the default is false - and unless explicitly set to true by either strategy or config - it'll remain at false - hence it'll trail immediately - with every point the trade reaches a new high.

**Comment by Vaadasch** [comment]
Ok, thanks a lot, i've got it now. 
*Sigh*, that is asking some gymnastic to understand when we see it for the first time.
I'll see if i can update the docs to enlight this difference.

**Comment by Vaadasch** [comment]
It is not much, but i hope it will suffice to no longer bother you on this point :-) 
#11306
